src/ForwardedHttpMiddleware.cs

```diff
@@ -1,27 +1,885 @@
+        {
+            if (_options.AllowedHosts == null || _options.AllowedHosts.Count == 0)
+            {
+                _allowAllHosts = true;
+                return;
+            }
+
+            var allowedHosts = new List<StringSegment>();
+            foreach (var entry in _options.AllowedHosts)
+            {
+                // Punycode. Http.Sys requires you to register Unicode hosts, but the headers contain punycode.
+                var host = new HostString(entry).ToUriComponent();
+
+                if (IsTopLevelWildcard(host))
+                {
+                    // Disable filtering
+                    _allowAllHosts = true;
+                    return;
+                }
+
+                if (!allowedHosts.Contains(host, StringSegmentComparer.OrdinalIgnoreCase))
+                {
+                    allowedHosts.Add(host);
+                }
+            }
+
+            _allowedHosts = allowedHosts;
+        }
+
+        private bool IsTopLevelWildcard(string host)
+        {
+            return (string.Equals("*", host, StringComparison.Ordinal) // HttpSys wildcard
+                           || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
+                           || string.Equals("0.0.0.0", host, StringComparison.Ordinal)); // IPv4 Any
         }
 
         public Task Invoke(HttpContext context)
         {
+            ApplyForwarders(context);
             return _next(context);
         }
+
+        private void ApplyForwarders(HttpContext context)
+        {
+            StoreOriginalValues(context);
+
+            var checkFor = _options.ForwardedHttp.HasFlag(ForwardedHttp.For);
+            var checkBy = _options.ForwardedHttp.HasFlag(ForwardedHttp.By);
+            var checkProto = _options.ForwardedHttp.HasFlag(ForwardedHttp.Proto);
+            var checkHost = _options.ForwardedHttp.HasFlag(ForwardedHttp.Host);
+
+            var request = context.Request;
+            var forwarded = request.Headers[ForwardedHeader];
+            var success = ParseProxyValues.TryParse(forwarded, out var forwardedHeader);
+            if (!success)
+            {
+                _logger.LogWarning("Invalid forwarded header");
+                return;
+            }
+
+            var entryCount = forwardedHeader.ForwardedValues.Count;
+
+            // Apply ForwardLimit, if any
+            if (_options.ForwardLimit.HasValue && entryCount > _options.ForwardLimit)
+            {
+                entryCount = _options.ForwardLimit.Value;
+            }
+
+            var sets = forwardedHeader.ForwardedValues
+                .Select(x => new SetOfForwarders
+                {
+                    RemoteIpAndPortText = x.For,
+                    LocalIpAndPortText = x.By,
+                    Scheme = x.Proto,
+                    Host = x.Host
+                })
+                .Reverse()
+                .Take(entryCount)
+                .ToArray();
```

Comment:
Scope the list before allocating.
```suggestion
            var sets = forwardedHeader.ForwardedValues
                .Reverse()
                .Take(entryCount)
                .Select(x => new SetOfForwarders
                {
                    RemoteIpAndPortText = x.For,
                    LocalIpAndPortText = x.By,
                    Scheme = x.Proto,
                    Host = x.Host
                })
                .ToArray();
```